src/components/tl/ucp/alltoall/alltoall_bruck.c

```diff
@@ -165,29 +188,89 @@ void ucc_tl_ucp_alltoall_bruck_progress(ucc_coll_task_t *coll_task)
         step = 1 << (task->alltoall_bruck.iteration - 1);
     }
 
-    st = ucc_mc_memcpy(PTR_OFFSET(args->dst.info.buffer, trank * seg_size),
-                       PTR_OFFSET(args->src.info.buffer, trank * seg_size),
+    st = ucc_mc_memcpy(PTR_OFFSET(task->alltoall_bruck.dst, trank * seg_size),
+                       PTR_OFFSET(task->alltoall_bruck.src, trank * seg_size),
                        seg_size, UCC_MEMORY_TYPE_HOST, UCC_MEMORY_TYPE_HOST);
     if (ucc_unlikely(st != UCC_OK)) {
         task->super.status = st;
         return;
     }
-    task->super.status =
-        ucc_tl_ucp_alltoall_bruck_backward_rotation(args->dst.info.buffer,
-                                                    scratch, trank, tsize,
-                                                    seg_size);
+    st = ucc_tl_ucp_alltoall_bruck_backward_rotation(task->alltoall_bruck.dst,
+                                                     scratch, trank, tsize,
+                                                     seg_size);
+    if (ucc_unlikely(st != UCC_OK)) {
+        task->super.status = st;
+        return;
+    }
+
+    if (smtype != UCC_MEMORY_TYPE_HOST || dmtype != UCC_MEMORY_TYPE_HOST) {
+        task->alltoall_bruck.phase = PHASE_BCOPY;
+        st = ucc_coll_task_get_executor(&task->super, &exec);
+        if (ucc_unlikely(st != UCC_OK)) {
+            task->super.status = st;
+            return;
+        }
+
+        eargs.task_type = UCC_EE_EXECUTOR_TASK_COPY;
+        eargs.copy.src  = task->alltoall_bruck.dst;
+        eargs.copy.dst  = args->dst.info.buffer;
+        eargs.copy.len  = seg_size * tsize;
+        st = ucc_ee_executor_task_post(exec, &eargs,
+                                       &task->alltoall_bruck.etask);
+        if (ucc_unlikely(st != UCC_OK)) {
+            task->super.status = st;
+            return;
+        }
+        st = ucc_ee_executor_task_test(task->alltoall_bruck.etask);
```

Comment:
test it her just in case copy completes immediately so we can mark collective as completed right away. I added EXEC_TASK_TEST macro to avoid code duplication